### PR TITLE
Add a config for automatically including `nonce` in `javascript_tag`, `javascript_include_tag` and `stylesheet_link_tag`

### PR DESCRIPTION
**Pull request by francktrouillez**
### Motivation / Background

When using `nonce` in context of [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP), the `nonce` attribute can currently not be added automatically to the tags used in the views.

This means that the `nonce` attribute must be added manually to each tag (`javascript_tag`, `javascript_include_tag` and `stylesheet_link_tag`) that requires it.

From my understanding, there's currently no way to automatically add those nonces to the tags.

The current solutions to this I could identify are:
- Overriding the `javascript_include_tag` and `stylesheet_link_tag` helpers to explicitly add the `nonce` attribute.
- Adding manually `nonce: true` to each tag, even though this means tags in views coming straight from third-parties will not benefit from it, and might be blocked by the CSP, even though my application should trust them.

Ideally, I would like to have access to a configuration option to automatically enable the `nonce` attribute on the tags that require it.

I don't know if there's a better way to handle this, or if there's a specific reason why we wouldn't want to.

### Detail

This pull request main change is the addition of a new configuration option `content_security_policy_nonce_auto` that allows to automatically include a `nonce` to tags affected by the directives specified in the config option `content_security_policy_nonce_directives`. In order to keep the current behavior, the new configuration option is set to `false` by default.

### Additional information

This would allow developers to easily enable the `nonce` attribute for the tags that require it, without having to manually add it to each tag. This also means that gems that are trusted and used by the application would automatically benefit from the `nonce` attribute, even if the views are internal to the third-party.

Once again, I don't know if there's a better way to handle this, or if there's a specific reason why we wouldn't want to allow such behavior.

I'm open to any feedback or suggestions on how to improve this pull request.

Thank you for your time and consideration!

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by francktrouillez** [comment]
Thanks @byroot for the quick answer!

However, I'm not sure #53794 aimed at solving the same issue as this PR does.

The issue I'm trying to solve is that when using `nonce`, the `nonce` attribute can currently not be added automatically to the tags used in the views, meaning I need to specify manually `nonce: true` to each tag that requires it.

For instance, this test wouldn't pass without the `nonce: true` attribute specified for `javascript_include_tag` and `stylesheet_link_tag`:

https://github.com/rails/rails/blob/59566cab94ca2f5c43147117870f0ee0ff73fed3/actionview/test/template/asset_tag_helper_test.rb#L727-L736

From my understanding, #53794 aimed at adding the nonce for preloaded links, which I believe is a different issue.

Am I missing something? Or is there a specific reason why we wouldn't want to allow automatic `nonce` addition?

Thank you very much!

File: railties/lib/rails/generators/rails/app/templates/config/initializers/content_security_policy.rb.tt [hunk 1]
```diff
@@ -20,6 +20,9 @@
 #   config.content_security_policy_nonce_generator = ->(request) { request.session.id.to_s }
 #   config.content_security_policy_nonce_directives = %w(script-src style-src)
 #
+#   # Automatically add `nonce` to permitted directives.
+#   # config.content_security_policy_nonce_auto = true
```
Comment (byroot) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/content_security_policy.rb.tt:24]: Is there a reason you'd want this to false? Because I'm thinking it's yet another extra config that I'd rather avoid.

Comment (francktrouillez) [reply railties/lib/rails/generators/rails/app/templates/config/initializers/content_security_policy.rb.tt:24]: To be honest, I was trying to stay as retro-compatible as possible.

Now, if we put that aside, I believe there could be applications in which you actually want to "choose" what scripts are allowed to run (which is the current behavior).

What do you think?

_Note_: Personally, I think I'd always have that set to `true` in my projects, but I didn't want to enforce it for everyone by default.

Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/config/initializers/content_security_policy.rb.tt:24]: Ok, let me rephrase. Is there a case where you have a CSP nonce directive but wouldn't want to set the nonce on a script or style tag?

Comment (byroot) [reply railties/lib/rails/generators/rails/app/templates/config/initializers/content_security_policy.rb.tt:24]: I don't have that much thoughts, it's been a while since I worked on frontend much, so I don't have a lot of experience with CSP. I might defer to someone else.

Comment (francktrouillez) [reply railties/lib/rails/generators/rails/app/templates/config/initializers/content_security_policy.rb.tt:24]: @byroot

Did you get the chance to take a second look over this? Or do you think involving someone else might be relevant in this case?

Let me know if I can do something else to help with this!

**Comment by francktrouillez** [comment]
Is there anything else I can do to help with the progress of this PR? @byroot

**Comment by byroot** [comment]
@francktrouillez I'm sorry, I asked for other committers opinions but got crickets.

@flavorjones are you familiar with CSP? If so what's you opinion on the PR, seems sensible to me but my frontend knowledge is very outdated.

**Comment by intrip** [comment]
I don't think we should add a nonce by default to `javascript_include_tag` or `stylesheet_link_tag` directives because, whilst is convenient, it discourages the usage of [host source](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#host-source), which I believe is more appropriate (and secure) for this use-case. @francktrouillez can you elaborate further on why you don't want to use the `host-source` instead?

On the other hand, it make sense adding a nonce by default for the `javascript_tag` directives, it is used for inline scripts and in this case, you can't rely on the `host-source` directive; but we also need to verify that `content_security_policy_nonce_generator` is also configured before doing so, otherwise this would allow an easy CSP bypass. Finally, we should also cover the `tag.script` method call, which can also be used for inline scripts.

**Comment by francktrouillez** [comment]
Thanks @intrip for your feedback!

> I don't think we should add a `nonce` by default to `javascript_include_tag` or `stylesheet_link_tag` directives because, whilst is convenient, it discourages the usage of [host source](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#host-source), which I believe is more appropriate (and secure) for this use-case. @francktrouillez can you elaborate further on why you don't want to use the `host-source` instead?

Indeed, I would say host sources are more straightforward than using nonces. The main reason I believe it makes sense to still enable it for `javascript_include_tag` and `stylesheet_link_tag` is that the use of
[`strict-dynamic`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#strict-dynamic) is very convenient for integrating third-parties, even though it disables host sources, in order to only use sources that are fingerprinted with a nonce or a hash. It allows to basically trust a script that is loaded from a trusted source (using a `nonce`), and then let it load other scripts without having to explicitly add them to the CSP.

In case someone uses `strict-dynamic`, there's AFAIK no way to automatically add a `nonce` to the `javascript_include_tag` and `stylesheet_link_tag` directives, either in the source code (can be managed by copy-pasting the line everywhere) or in the **third-party code** (harder to manage without requiring the third-party to add it explicitly, or without overriding the current helper methods).

I personally don't have a strong opinion whether this is good practice or not to have `strict-dynamic`, but the fact that this exists, and that Rails supports it, makes me think that it would be a good idea to improve its usage in Rails as well.

What do you think?

> On the other hand, it make sense adding a `nonce` by default for the `javascript_tag` directives, it is used for inline scripts and in this case, you can't rely on the `host-source`; but we also need to verify that `content_security_policy_nonce_generator` is also configured before doing so, otherwise this would allow an easy CSP bypass. Finally, we should also cover the `tag.script` method call, which can also be used for inline scripts.

I agree with you for the `content_security_policy_nonce_generator` presence check, thanks for pointing it out! I already made the change in the PR.

For the `tag.script`, do you think it makes sense to automatically add it there as well? I personally see the `tag` helper method more as a raw way to add an HTML element, while `javascript_tag` is a more elaborate wrapper that could indeed implement such feature. For instance, I see `tag` currently doesn't support `nonce: true`, while `javascript_tag`, `javascript_include_tag` and `stylesheet_link_tag` do. I don't have a strong opinion on this though, but I'm usually not a huge fan of adding exceptions to such generic helper, as it usually makes it harder to understand and maintain.

What do you think?

**Comment by intrip** [comment]
> In case someone uses strict-dynamic, there's AFAIK no way to automatically add a nonce to the javascript_include_tag and stylesheet_link_tag directives, either in the source code (can be managed by copy-pasting the line everywhere) or in the third-party code (harder to manage without requiring the third-party to add it explicitly, or without overriding the current helper methods).

Ah, I see what you mean: If you use `strict-dynamic` you're forced to set a nonce. Personally, I don't think this a good reason to add an option that will auto add a nonce to all `javascript_include_tag `, `stylesheet_link_tag `, and `javascript_tag `. I suggest a solution based on having a good default that can be overridden on a per-tag basis if desired. 

Since we know that:
- If you use `javascript_tag` you need a nonce
- if you use `strict-dynamic` you need a nonce

Maybe we can automatically add the nonce in both cases without adding any additional configuration option? This would change the current behaviour, but I don't think it'll add any compatibility issue. And if someone doesn't want the nonce is still possible to override the default when calling the method `javascript_include_tag(..., nonce: false)`

> For the tag.script, do you think it makes sense to automatically add it there as well? I personally see the tag helper method more as a raw way to add an HTML element, while javascript_tag is a more elaborate wrapper that could indeed implement such feature. For instance, I see tag currently doesn't support nonce: true, while javascript_tag, javascript_include_tag and stylesheet_link_tag do. I don't have a strong opinion on this though, but I'm usually not a huge fan of adding exceptions to such generic helper, as it usually makes it harder to understand and maintain.

Right! Agree on not touching the `tag` helper 👍

**Comment by francktrouillez** [comment]
@intrip 
> I suggest a solution based on having a good default that can be overridden on a per-tag basis if desired.
>
> Since we know that:
>
> * If you use `javascript_tag` you need a nonce
> * if you use `strict-dynamic` you need a nonce
>
> Maybe we can automatically add the nonce in both cases without adding any additional configuration option? This would change the current behaviour, but I don't think it'll add any compatibility issue. And if someone doesn't want the nonce is still possible to override the default when calling the method `javascript_include_tag(..., nonce: false)`

So, based on the current changes proposed in the PR, and if I understand correctly, you would remove the configuration option and always add the `nonce` to those tags if the corresponding directive is present in the `content_security_policy_nonce_directives`. Is that correct?

It makes sense to me.

My only concern is that this change will not keep the current behavior, as it changes the default behavior of the `javascript_tag`, `javascript_include_tag` and `stylesheet_link_tag` helpers. That's why I went for the configuration option set to `false` by default in the first place, but as you said, I don't expect it to break applications. I might be missing something though, and that's why I wanted to go the safe road with the configuration option, that can switch its default value to `true` in future releases.

What do you think?

_Note_: `nonce: false` is currently a bit misleading today, as it actually adds `nonce="false"` to the tag, instead of not adding it in the first place, which I believe is what would be expected. I'll work on this if this PR is merged, as I don't think it makes sense to specify `nonce: false` without the changes in this PR (you could have a niche case where we want to set the `nonce` dynamically though).

**Comment by dhh** [comment]
Appreciate the work on this, but I concur with @intrip. Another problem with inline nonces is that they don't work with partial caching. I think we're better off keeping the nonce usage explicit. This is especially true when you consider that plugins might use javascript_tag inside a cache block, and if suddenly that starts adding nonces, things are going to silently break.

**Comment by francktrouillez** [comment]
Thanks @dhh for the feedback, really appreciate it!

> Appreciate the work on this, but I concur with @intrip. Another problem with inline nonces is that they don't work with partial caching. I think we're better off keeping the nonce usage explicit. This is especially true when you consider that plugins might use javascript_tag inside a cache block, and if suddenly that starts adding nonces, things are going to silently break.

I understand your concerns, but I still feel that in the current state of Rails, the use case I described can't really be filled.

Today, integrating `strict-dynamic` source expression in Rails is challenging in scenarios where third-party views are involved. This is because these parts of the application may not be set up to receive or propagate the nonce, causing CSP to block/report the scripts.

AFAIK, `strict-dynamic` makes the `script-src` directive ignore every other source expression (except nonces and hashes obviously). Knowing that, I can't really use `strict-dynamic` with 3rd-parties views, as I can't explicitly add the `nonce` in the 3rd-party script tags. The only way I could find today was to **override** the methods that generate those tags, like `javascript_include_tag`.

The fact that the only way to use `strict-dynamic` is to override the Rails methods that generate the tags makes me think that this should be supported by Rails.

Based on that, I thought about adding automatically a `nonce`, as to solve the use case of using `strict-dynamic` with 3rd-party views. At the same time, I didn't want to break the current behavior of the methods that generate the tags. Like you mentioned, this could silently break in case partial caching is involved. That's why I went with an approach of adding a configuration to enable the `nonce` automatically, that defaults to `false`, as to **stay retro-compatible**. If the developer wants to use `strict-dynamic`, they can enable the configuration and use it, but by default, the **exact same behavior remains**.

So, we wouldn't add the `nonce` automatically by default, but we would allow the developer to enable the configuration if they want to use `strict-dynamic`.

I might be missing something, but I don't see how this approach can break current applications as either:
- They already use `strict-dynamic`, and have a tweaked way of adding the `nonce` to the tags (handling partial caching, etc.); or
- They don't use `strict-dynamic`, and the behavior remains the same

Something I might be missing is that maybe **Rails took position** on the use of `strict-dynamic` and doesn't want to have more support for it. In that case, I would understand that such changes are not the way to go. Personally, the fact that this source expression exists makes me think it would be great to support it, the same way Rails does with the rest of them.

Please, let me know if I'm missing something, or if you have any other concerns about this approach.
I'm open to any feedback!

**Comment by dhh** [comment]
You make a fair point with 3rd party views that you can't control but that you can vet. I'm convinced.

But we need to spell out the consequences, like with caching, in the place where you've added the invitation to use this flag. So people know what the ramifications might be, and also why this is even there (primarily to deal with 3rd party views).

**Comment by francktrouillez** [comment]
Thanks @dhh for the quick reaction!

> we need to spell out the consequences, like with caching, in the place where you've added the invitation to use this flag. So people know what the ramifications might be, and also why this is even there (primarily to deal with 3rd party views).

I made changes to the PR to take into account your last feedback.

Let me know if that works for you, or if you have any other feedback!

File: actionview/lib/action_view/helpers/asset_tag_helper.rb [hunk 2]
```diff
@@ -135,7 +137,7 @@ def javascript_include_tag(*sources)
             "src" => href,
             "crossorigin" => crossorigin
           }.merge!(options)
-          if tag_options["nonce"] == true
+          if tag_options["nonce"] == true || (!tag_options.key?("nonce") && auto_include_nonce_for_scripts)
```
Comment (justinko) [inline actionview/lib/action_view/helpers/asset_tag_helper.rb:140]: `tag_options.exclude?("nonce")`
